Transient IT failure during managed_access_repository_sources

https://gitlab.anvil.gi.ucsc.edu/ucsc/azul/-/jobs/38275

```
2024-04-16 04:40:07,494    INFO MainThread test.integration_test: Beginning sub-test [managed_access_repository_sources] {'catalog': 'anvil-it'}
2024-04-16 04:40:07,496    INFO MainThread azul.oauth2: Making GET request to 'https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it'
2024-04-16 04:40:07,496   DEBUG MainThread azul.oauth2: … without request body
2024-04-16 04:40:08,018    INFO MainThread azul.oauth2: Got 200 response after 0.522s from GET to https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it
2024-04-16 04:40:08,018   DEBUG MainThread azul.oauth2: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '469', 'Connection': 'keep-alive', 'Date': 'Tue, 16 Apr 2024 04:40:08 GMT', 'x-amzn-RequestId': 'd2ae6113-822d-4328-929d-788703112cc9', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'x-amz-apigw-id': 'WTUeNHSRIAMEHhA=', 'X-Content-Type-Options': 'nosniff', 'X-Amzn-Trace-Id': 'Root=1-661e0127-1a65f82f5a41699639ef7837;Parent=67f48baca56efb4f;Sampled=0;lineage=87986af1:0', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 0a2ddb6f9b0df10d973faa154be16dba.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': '9Nc5KhUGeCf_3HmZJkJxmq6FmzOLg01mOEkOerMC8KCx9Z2Fx9H66Q=='})
2024-04-16 04:40:08,019   DEBUG MainThread azul.oauth2: … with response body b'{"sources":[{"sourceId":"7d632e2a-54f1-4021-81fd-dbaf893715cd","sourceSpec":"tdr:datarepo-dev-97ad270b:snapshot/ANVIL_CMG_Sam...'
2024-04-16 04:40:08,020    INFO MainThread azul.oauth2: Making GET request to 'https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it'
2024-04-16 04:40:08,020   DEBUG MainThread azul.oauth2: … without request body
2024-04-16 04:40:09,075    INFO MainThread azul.oauth2: Got 200 response after 1.056s from GET to https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it
2024-04-16 04:40:09,075   DEBUG MainThread azul.oauth2: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '166', 'Connection': 'keep-alive', 'Date': 'Tue, 16 Apr 2024 04:40:09 GMT', 'x-amzn-RequestId': 'fc45c15c-eec7-44f8-b8fd-0c24ffd98b63', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'x-amz-apigw-id': 'WTUeTFGXoAMErQQ=', 'X-Content-Type-Options': 'nosniff', 'X-Amzn-Trace-Id': 'Root=1-661e0128-26fe1d11049d4eee6ec33bcf;Parent=74e0e39690b2f4fb;Sampled=0;lineage=87986af1:0', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 d0e0fdfe87d75193de6278b5eca393f8.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': 'XNuYKs5PXUbLBrRkQbsArhD0hP-fTzuyR6kEb5h6a7KPVzKN0YnpWg=='})
2024-04-16 04:40:09,076   DEBUG MainThread azul.oauth2: … with response body b'{"sources":[{"sourceId":"790795c4-49b1-4ac8-a060-207b92ea08c5","sourceSpec":"tdr:datarepo-dev-e53e74aa:snapshot/ANVIL_1000G_2...'
2024-04-16 04:40:09,294    INFO MainThread azul.terra: Making GET request to 'https://sam.dsde-dev.broadinstitute.org/register/user/v1'
2024-04-16 04:40:09,294   DEBUG MainThread azul.terra: … without request body
2024-04-16 04:40:09,498    INFO MainThread azul.terra: Got 404 response after 0.204s from GET to https://sam.dsde-dev.broadinstitute.org/register/user/v1
2024-04-16 04:40:09,498   DEBUG MainThread azul.terra: … with response headers HTTPHeaderDict({'Date': 'Tue, 16 Apr 2024 04:40:09 GMT', 'Server': 'akka-http/10.2.9', 'X-Frame-Options': 'SAMEORIGIN', 'X-XSS-Protection': '1; mode=block', 'X-Content-Type-Options': 'nosniff', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Referrer-Policy': 'strict-origin-when-cross-origin', 'Access-Control-Allow-Origin': '*', 'Access-Control-Allow-Headers': 'authorization,content-type,accept,origin,x-app-id', 'Access-Control-Allow-Methods': 'GET,POST,PUT,PATCH,DELETE,OPTIONS,HEAD', 'Access-Control-Max-Age': '1728000', 'Content-Type': 'application/json', 'Content-Length': '122', 'Via': '1.1 google', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'})
2024-04-16 04:40:09,499   DEBUG MainThread azul.terra: … with response body b'{"causes":[],"message":"Google Id 110148757117121773195 not found in sam","source":"sam","stackTrace":[],"statusCode":404}'
2024-04-16 04:40:09,499    INFO MainThread azul.terra: Making GET request to 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots?offset=0&limit=1000&sort=created_date&direction=asc'
2024-04-16 04:40:09,499   DEBUG MainThread azul.terra: … without request body
2024-04-16 04:40:09,916    INFO MainThread azul.terra: Got 403 response after 0.417s from GET to https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots?offset=0&limit=1000&sort=created_date&direction=asc
2024-04-16 04:40:09,916   DEBUG MainThread azul.terra: … with response headers HTTPHeaderDict({'Date': 'Tue, 16 Apr 2024 04:40:09 GMT', 'Server': 'Apache', 'X-Frame-Options': 'SAMEORIGIN', 'Access-Control-Allow-Headers': 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range,Authorization,Accept,Referer,X-App-Id,Origin', 'Access-Control-Allow-Methods': 'GET,POST,DELETE,PUT,PATCH,OPTIONS,HEAD', 'X-Content-Type-Options': 'nosniff', 'Strict-Transport-Security': 'max-age=31536000;includeSubDomains', 'Cache-Control': 'no-cache,no-store,must-revalidate', 'X-Request-ID': 'e74nXGzy', 'Content-Type': 'application/json', 'Content-Length': '79', 'Vary': 'Origin', 'Via': '1.1 google', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000'})
2024-04-16 04:40:09,916   DEBUG MainThread azul.terra: … with response body b'{"message":"Google Id 110148757117121773195 not found in sam","errorDetail":[]}'
2024-04-16 04:40:09,917    INFO MainThread azul.oauth2: Making GET request to 'https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it'
2024-04-16 04:40:09,917   DEBUG MainThread azul.oauth2: … without request body
2024-04-16 04:40:12,144    INFO MainThread azul.oauth2: Got 200 response after 2.227s from GET to https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it
2024-04-16 04:40:12,144   DEBUG MainThread azul.oauth2: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '166', 'Connection': 'keep-alive', 'Date': 'Tue, 16 Apr 2024 04:40:12 GMT', 'x-amzn-RequestId': 'dc418d99-3c4f-45e8-a60e-bc567a9f577d', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains', 'Access-Control-Allow-Headers': 'Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key', 'X-Frame-Options': 'DENY', 'x-amz-apigw-id': 'WTUemF-6IAMEltg=', 'X-Content-Type-Options': 'nosniff', 'X-Amzn-Trace-Id': 'Root=1-661e0129-4c411c0a68ae25ea2cec8e34;Parent=02d738b11b6fdb27;Sampled=0;lineage=87986af1:0', 'X-Cache': 'Miss from cloudfront', 'Via': '1.1 cfbbd45bc29ebb0e8475ffa26dff7618.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': '5pQCh_pTG94Z17uIYtfmOOnrbOXWY-uANZGLvjMrieJDvc_HvNA6Ow=='})
2024-04-16 04:40:12,144   DEBUG MainThread azul.oauth2: … with response body b'{"sources":[{"sourceId":"790795c4-49b1-4ac8-a060-207b92ea08c5","sourceSpec":"tdr:datarepo-dev-e53e74aa:snapshot/ANVIL_1000G_2...'
2024-04-16 04:40:12,145    INFO MainThread azul.terra: Making GET request to 'https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=foo'
2024-04-16 04:40:12,145   DEBUG MainThread azul.terra: … without request body
2024-04-16 04:40:12,198    INFO MainThread azul.terra: Got 400 response after 0.053s from GET to https://www.googleapis.com/oauth2/v3/tokeninfo?access_token=foo
2024-04-16 04:40:12,198   DEBUG MainThread azul.terra: … with response headers HTTPHeaderDict({'Expires': 'Mon, 01 Jan 1990 00:00:00 GMT', 'Pragma': 'no-cache', 'Date': 'Tue, 16 Apr 2024 04:40:12 GMT', 'Cache-Control': 'no-cache, no-store, max-age=0, must-revalidate', 'Content-Type': 'application/json; charset=UTF-8', 'Vary': 'X-Origin, Referer, Origin,Accept-Encoding', 'Server': 'ESF', 'X-XSS-Protection': '0', 'X-Frame-Options': 'SAMEORIGIN', 'X-Content-Type-Options': 'nosniff', 'Alt-Svc': 'h3=":443"; ma=2592000,h3-29=":443"; ma=2592000', 'Accept-Ranges': 'none', 'Transfer-Encoding': 'chunked'})
2024-04-16 04:40:12,198   DEBUG MainThread azul.terra: … with response body b'{\n  "error_description": "Invalid Value"\n}\n'
2024-04-16 04:40:12,199 WARNING MainThread azul.terra: Invalid credentials
Traceback (most recent call last):
  File "/builds/ucsc/azul/src/azul/terra.py", line 630, in for_registered_user
    self.validate()
  File "/builds/ucsc/azul/src/azul/oauth2.py", line 142, in validate
    reject(response.status == 400,
  File "/builds/ucsc/azul/src/azul/__init__.py", line 1544, in reject
    raise exception(*args)
azul.RequirementError: The token is not valid
2024-04-16 04:40:12,201    INFO MainThread azul.oauth2: Making GET request to 'https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it'
2024-04-16 04:40:12,202   DEBUG MainThread azul.oauth2: … without request body
2024-04-16 04:40:41,236    INFO MainThread azul.oauth2: Got 504 response after 29.034s from GET to https://service.anvil.gi.ucsc.edu/repository/sources?catalog=anvil-it
2024-04-16 04:40:41,236   DEBUG MainThread azul.oauth2: … with response headers HTTPHeaderDict({'Content-Type': 'application/json', 'Content-Length': '41', 'Connection': 'keep-alive', 'Date': 'Tue, 16 Apr 2024 04:40:41 GMT', 'x-amzn-RequestId': '94caf99c-dd3d-45e5-bdbf-2385f72b3209', 'x-amzn-ErrorType': 'InternalServerErrorException', 'x-amz-apigw-id': 'WTUe9H3uoAMEZiA=', 'X-Cache': 'Error from cloudfront', 'Via': '1.1 ff2d6deff1b50282a21f4b199088c76e.cloudfront.net (CloudFront)', 'X-Amz-Cf-Pop': 'IAD61-P3', 'X-Amz-Cf-Id': '8niMBU32CZ0slvOb_o9WPN7IZfDRiMLuunch8XysGo2KmnXyaidrjw=='})
2024-04-16 04:40:41,236   DEBUG MainThread azul.oauth2: … with response body b'{"message": "Endpoint request timed out"}'
2024-04-16 04:40:41,236    INFO MainThread test.integration_test: Failed sub-test [managed_access_repository_sources] {'catalog': 'anvil-it'}

…

======================================================================
FAIL: test_indexing (integration_test.IndexingIntegrationTest.test_indexing) [managed_access_repository_sources] (catalog='anvil-it')
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builds/ucsc/azul/test/integration_test.py", line 428, in subTest
    yield
  File "/builds/ucsc/azul/test/integration_test.py", line 1322, in _test_managed_access
    public_source_ids = self._test_managed_access_repository_sources(catalog,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/builds/ucsc/azul/test/integration_test.py", line 1357, in _test_managed_access_repository_sources
    self.assertEqual(401, self._get_url_unchecked(GET, url).status)
AssertionError: 401 != 504
----------------------------------------------------------------------
```

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -925,7 +925,7 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
                                 timeout=float(config.api_gateway_lambda_timeout + 1),
                                 retries=urllib3.Retry(total=5,
                                                       redirect=0,
-                                                      status_forcelist={429, 500, 502, 503}),
+                                                      status_forcelist={429, 500, 502, 503, 504}),
                                 redirect=False,
                                 preload_content=not stream)
         assert isinstance(response, urllib3.HTTPResponse)

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -413,8 +413,6 @@ class IntegrationTestCase(AzulTestCase, metaclass=ABCMeta):
 
 
 class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
-    num_fastq_bytes = 1024 * 1024
-
     _http: urllib3.request.RequestMethods
     _plain_http: urllib3.request.RequestMethods
 
@@ -1229,6 +1227,8 @@ class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
         content = BytesIO()
         storage_client.download_blob_to_file(str(url), content, start=0, end=size)
         return content
+
+    num_fastq_bytes = 1024 * 1024
 
     def _validate_fastq_content(self, content: ReadableFileObject):
         # Check signature of FASTQ file.

--- a/test/integration_test.py
+++ b/test/integration_test.py
@@ -413,13 +413,30 @@ class IntegrationTestCase(AzulTestCase, metaclass=ABCMeta):
 
 
 class IndexingIntegrationTest(IntegrationTestCase, AlwaysTearDownTestCase):
-    _http: urllib3.request.RequestMethods
+    """
+    An integration test case that tests indexing of public and managed-access
+    metadata from a random selection of bundles, and the expected effects on the
+    service API. This is our main integration test case.
+    """
+
+    #: A vanilla urllib3 HTTP client without authentication or any of the
+    #: special retry behaviour that we employ for Terra services. Note that
+    #: IT-specific retries are configured explicitly for each request, no matter
+    #: which client is used, in the :py:meth:`_get_url_unchecked` method.
+    #:
     _plain_http: urllib3.request.RequestMethods
+
+    #: Depending on the authorization context, this is either the same client as
+    #: the one refered to by the attribute above, or a client that sends an
+    #: access token — whose access token also depends on the context. Note that
+    #: IT-specific retries are configured explicitly for each request, no matter
+    #: which client is used, in the :py:meth:`_get_url_unchecked` method.
+    #:
+    _http: urllib3.request.RequestMethods
 
     def setUp(self) -> None:
         super().setUp()
         self._plain_http = http_client(log)
-        # Note that this attribute is swizzled in self._authorization_context
         self._http = self._plain_http
 
     @contextmanager

